platform/security/idp/security-idp-client/src/main/java/org/codice/ddf/security/idp/client/IdpMetadata.java

```diff
@@ -161,9 +178,45 @@ private EntityDescriptor getEntityDescriptor() {
       boolean updated = entryDescriptions.compareAndSet(null, edMap);
       if (!updated) {
         LOGGER.debug("Safe but concurrent update to serviceProviders map; using processed value");
+      } else {
+        setCacheInformation(extractRootEntityFromMap(edMap));
+      }
+    }
+
+    return extractRootEntityFromMap(edMap);
+  }
+
+  private void setCacheInformation(EntityDescriptor entityDescriptor) {
+    if (entityDescriptor != null) {
+      created = Instant.now();
+      if (entityDescriptor.getCacheDuration() != null) {
+        cacheDuration = Duration.ofMillis(entityDescriptor.getCacheDuration());
+      }
+      if (entityDescriptor.getValidUntil() != null) {
+        validUntil = entityDescriptor.getValidUntil().toDate().toInstant();
       }
     }
+  }
+
+  /**
+   * Return true if this the metadata is expired. It if ambiguous is both the cache duration and
+   * valid-until are specified. Assume that only one will be specified. Return true only if neither
+   * are currently valid. If neither is present, assume the metadata is good.
+   *
+   * @return true if metadata should be reacquired from the IDP
+   */
+  public boolean isMetadataExpired() {
+    boolean missingCacheDuration = cacheDuration == null;
+    boolean missingValidUntil = validUntil == null;
+    boolean missingAllTimeConstraints = missingCacheDuration && missingValidUntil;
+    Instant now = Instant.now();
+    boolean cacheDurationGood = !missingCacheDuration && now.isBefore(created.plus(cacheDuration));
+    boolean stillValidUntil = !missingValidUntil && now.isBefore(validUntil);
+    boolean isMetadataGood = missingAllTimeConstraints || cacheDurationGood || stillValidUntil;
+    return !isMetadataGood;
```

Comment:
I'm not sure the logic in here is necessarily correct. According to the spec:
>Note that cache expiration does not imply a lack of validity in the absence of a `validUntil` attribute or 
other information; failure to update a cached instance (e.g., due to network failure) need not render 
metadata invalid, although implementations may offer such controls to deployers.
[Section 4.3.1](https://www.oasis-open.org/committees/download.php/56786/sstc-saml-metadata-errata-2.0-wd-05-diff.pdf)

>Metadata MUST be considered invalid upon reaching the time specified in a `validUntil` attribute of the 
subject element(s). The effective expiration may be adjusted downward by parent element(s) with earlier 
expirations. Invalid metadata MUST NOT be used. This contrasts with "stale" metadata that may be beyond its optimum cache duration but is not explicitly invalid. Such metadata remains valid and MAY be 
used at the discretion of the implementation.
[Section 4.3.2](https://www.oasis-open.org/committees/download.php/56786/sstc-saml-metadata-errata-2.0-wd-05-diff.pdf)

I think this is stating if it valid until is there and it is invalid, we MUST expire the metadata regardless of if caching is good. I would say we should also add defensive logic if both are set to ensure that the `validUntil` is respected even if both happen to be set. 

And according to the first sentence in section 4.3.1 
>Document caching MUST be based on the duration indicated by the `cacheDuration` attribute

This implies that we **cannot** cache if the cacheDuration is not set and must  fetch new metadata every time. 

We could try to get @stustison or maybe @coyotesqrl to double check as well if we wanted.

Reply:
Looks like I was reading an outdated version of the spec. Yikes. :-(